src/test/txrequest_tests.cpp

```diff
@@ -0,0 +1,587 @@
+                    if (prio_prev <= prio_cur) {
+                        ok = false;
+                        break;
+                    }
+                }
+                if (!ok) break;
+            }
+            if (ok) {
+                ok = m_runner.txhashset.insert(ret).second;
+            }
+        } while(!ok);
+        return ret;
+    }
+
+    /** Generate a random GenTxid; the txhash follows NewTxHash; the is_wtxid flag is random. */
+    GenTxid NewGTxid(const std::vector<std::vector<NodeId>>& orders = {})
+    {
+        return {InsecureRandBool(), NewTxHash(orders)};
+    }
+
+    /** Generate a new random NodeId to use as peer. The same NodeId is never returned twice
+     *  (across all Scenarios combined). */
+    NodeId NewPeer()
+    {
+        bool ok;
+        NodeId ret;
+        do {
+            ret = InsecureRandBits(63);
+            ok = m_runner.peerset.insert(ret).second;
+        } while(!ok);
+        return ret;
+    }
+
+    std::chrono::microseconds Now() const { return m_now; }
+};
+
+/** Add to scenario a test with a single tx announced by a single peer.
+ *
+ * config is an integer between 0 and 31, which controls which variant of the test is used.
+ */
+void BuildSingleTest(Scenario& scenario, int config)
+{
+    auto peer = scenario.NewPeer();
+    auto gtxid = scenario.NewGTxid();
+    bool immediate = config & 1;
+    bool preferred = config & 2;
+    auto delay = immediate ? NO_TIME : RandomTime8s();
+
+    scenario.SetTestName(strprintf("Single(config=%i)", config));
+
+    // Receive an announcement, either immediately requestable or delayed.
+    scenario.ReceivedInv(peer, gtxid, preferred, immediate ? MIN_TIME : scenario.Now() + delay);
+    if (immediate) {
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s1");
+    } else {
+        scenario.Check(peer, {}, 1, 0, 0, "s2");
+        scenario.AdvanceTime(delay - MICROSECOND);
+        scenario.Check(peer, {}, 1, 0, 0, "s3");
+        scenario.AdvanceTime(MICROSECOND);
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s4");
+    }
+
+    if (config >> 3) { // We'll request the transaction
+        scenario.AdvanceTime(RandomTime8s());
+        auto expiry = RandomTime8s();
+        scenario.Check(peer, {gtxid}, 1, 0, 0, "s5");
+        scenario.RequestedTx(peer, gtxid.GetHash(), scenario.Now() + expiry);
+        scenario.Check(peer, {}, 0, 1, 0, "s6");
+
+        if ((config >> 3) == 1) { // The request will time out
+            scenario.AdvanceTime(expiry - MICROSECOND);
+            scenario.Check(peer, {}, 0, 1, 0, "s7");
+            scenario.AdvanceTime(MICROSECOND);
+            scenario.Check(peer, {}, 0, 0, 0, "s8");
+            return;
+        } else {
+            scenario.AdvanceTime(std::chrono::microseconds{InsecureRandRange(expiry.count())});
+            scenario.Check(peer, {}, 0, 1, 0, "s9");
+            if ((config >> 3) == 3) { // A reponse will arrive for the transaction
```

Comment:
ea7839c8889864be174b85a1f4de326f5e12e430, typo:
```suggestion
            if ((config >> 3) == 3) { // A response will arrive for the transaction
```

Reply:
Done.